add Yahoo currency provider

remedge opened the pull request:
- add YahooCurrencyRateProvider;
- add test for new provider;
- add guzzle dependency to composer.json

remedge committed:
add provider

- add YahooCurrencyRateProvider;
- add test for new provider;
- add guzzle dependency to composer.json

remedge committed:
replace guzzle with curl

maZahaca commented on src/RedCode/Currency/Rate/Provider/YahooCurrencyRateProvider.php:
```diff
@@ -0,0 +1,158 @@
+
+    public function __construct(ICurrencyRateManager $currencyRateManager, ICurrencyManager $currencyManager)
+    {
+        $this->currencyRateManager  = $currencyRateManager;
```
These properties must be defined in the class scope.
See details of inspection [here](https://scrutinizer-ci.com/g/redco/redcode-currency-rate/inspections/2f6f9ee3-a150-4a96-890f-bb06952f21d9)

maZahaca commented on src/RedCode/Currency/Rate/Provider/YahooCurrencyRateProvider.php:
```diff
@@ -0,0 +1,158 @@
+            $date = new \DateTime();
+            $results = $this->_processActualRequest($currencyCodes);
+        } else {
+            $results = $this->_processHistoricalRequest($currencyCodes, $date);
```
Is it really two different ways to retrieve current and past rates information?

maZahaca commented on src/RedCode/Currency/Rate/Provider/YahooCurrencyRateProvider.php:
```diff
@@ -0,0 +1,158 @@
+    }
+
+    /**
+     * @param $currencyCodes
```
specify param as array

maZahaca commented on src/RedCode/Currency/Rate/Provider/YahooCurrencyRateProvider.php:
```diff
@@ -0,0 +1,158 @@
+     *
+     * @return array
+     */
+    private function _processActualRequest($currencyCodes)
```
specify type hint as array

maZahaca commented on src/RedCode/Currency/Rate/Provider/YahooCurrencyRateProvider.php:
```diff
@@ -0,0 +1,158 @@
+        $requestString = '/v1/public/yql?q=';
+        $query = 'select * from yahoo.finance.xchange where pair in ("'.implode('","', $currencyCodes).'")';
+        $query .= '&env=store://datatables.org/alltableswithkeys';
+        $query = str_replace(' ', '%20', $query);
```
Why is that?

maZahaca commented on src/RedCode/Currency/Rate/Provider/YahooCurrencyRateProvider.php:
```diff
@@ -0,0 +1,158 @@
+     */
+    private function _processActualRequest($currencyCodes)
+    {
+        $baseUrl =  'http://query.yahooapis.com';
```
It's better to:
- build url params with [http_build_query](http://php.net/manual/en/function.http-build-query.php)
- move base url to the class constant

maZahaca commented on src/RedCode/Currency/Rate/Provider/YahooCurrencyRateProvider.php:
```diff
@@ -0,0 +1,158 @@
+    private function _processHistoricalRequest($currencyCodes, \DateTime $date)
+    {
+        $url = 'http://finance.yahoo.com/connection/currency-converter-cache';
+        $query = '?date=' . $date->format('Ymd');
```
The [same](https://github.com/redco/redcode-currency-rate/pull/12/files#r42325552)

maZahaca commented on src/RedCode/Currency/Rate/Provider/YahooCurrencyRateProvider.php:
```diff
@@ -0,0 +1,158 @@
+     */
+    private function _processHistoricalRequest($currencyCodes, \DateTime $date)
+    {
+        $url = 'http://finance.yahoo.com/connection/currency-converter-cache';
```
Please combine two requests to one,
See on [this example](https://developer.yahoo.com/yql/console/?q=show%20tables&env=store://datatables.org/alltableswithkeys#h=select+*+from+yahoo.finance.historicaldata+where+symbol+IN+%28%22EUR%3DX%22%2C+%22GBP%3DX%22%29+and+startDate+%3D+%222015-09-11%22+and+endDate+%3D+%222015-09-11%22) to get historical data
Use "Open" as a rate

remedge replied:
The problem is that yahoo.finance.historicaldata don't provide currencies info for Saturday and Sunday. And most important, it depends on US time zone. Due to the time difference, it will not provide information on the current day of at least 7 hours for Russia for example.

I suggest to use yahoo.finance.xchange for getting current day currencies. In another case, it is necessary to use some additional validation logic for date.

maZahaca replied:
What do you mean by additional validation logic for date?

remedge replied:
I mean check date for holiday and check date for existing in US timezone before sending to yahoo.finance.historicaldata.

Another variant - to check output from yahoo.finance.historicaldata requerst and if its empty, perform another request to yahoo.finance.xchange

maZahaca replied:
Ok, let's stay on the variant of usage yahoo.finance.historicaldata with dates, when it's possible. 
When it's NOT throw an exception NoRatesAvailableForDateException [date, provider]. Check [this file](https://github.com/redco/redcode-currency-rate/blob/master/src/RedCode/Currency/Rate/Exception/RateNotFoundException.php) for get an example.

It would be great if you can also change \Exception -> NoRatesAvailableForDateException
in [the file](https://github.com/redco/redcode-currency-rate/blob/master/src/RedCode/Currency/Rate/Provider/EcbCurrencyRateProvider.php#L48)

maZahaca replied:
So you mean yahoo.finance.xchange has the same issue about time zones?

remedge replied:
As i know yahoo.finance.xchange has no option for providing dates. Its return only current day values

maZahaca replied:
Yes, but I guess it also operates with US timezones inside.
Depends on service timezone, you need to detect which service to call.

remedge committed:
updates on comments

- add BadXMLQueryException;
- add NoRatesAvailableForDateException;
- update YahooCurrencyRateProvider to use only
yahoo.finance.historicaldata;
- add new exceptions to EcbCurrencyRateProvider

remedge committed:
update tests

remedge commented:
- YahooCurrencyProvider updated to use only yahoo.finance.historicaldata table;
- Exceptions NoRatesAvailableForDateException and BadXMLQueryException added and integrated to XML-using currency providers;
- XMLLoader added and integrated to YahooCurrencyProvider and EcbCurrencyProvider for optimize testing;
- Tests are updated.

maZahaca commented on src/RedCode/Currency/Rate/Exception/NoRatesAvailableForDateException.php:
```diff
@@ -0,0 +1,51 @@
+    /**
+     * @var \RedCode\Currency\ICurrency
+     */
+    protected $currency;
```
Copy-past?
No any usage of this in the file

remedge replied:
Obviously, yes :) fixed

remedge committed:
fix copy-past & test

maZahaca commented on tests/EcbCurrencyRateProviderTest.php:
```diff
@@ -87,7 +88,8 @@ function (ICurrency $currency, ICurrencyRateProvider $provider, \DateTime $date,
             $currencyRateManager,
-            $currencyManager
+            $currencyManager,
+            new XMLLoader()
```
You did excellent job about XMLLoader, this is actually I wanted to ask you to do.

Because it will improve tests.
And from now during the tests it's not necessary to request real providers for data.
You should mock XMLLoader::load method and return general XML and error XML to check some of your exceptions!

maZahaca replied:
Ouh, you've already added test for wrong answer. Add mocking for others

maZahaca replied:
The same as here https://github.com/redco/redcode-currency-rate/pull/12/files#r43361903
Use only mocking for XMLLoader in provider tests

maZahaca commented on tests/EcbCurrencyRateProviderTest.php:
```diff
@@ -157,5 +159,34 @@ public function testEcbCurrencyRateProviderGetRatesYesterday()
+        $currencyRateManager = $this->getMock('\\RedCode\\Currency\\Rate\\ICurrencyRateManager');
+
+        $xmlLoader = $this->getMock('\\RedCode\\Currency\\Rate\\XML\\XMLLoader');
+        $xmlLoader
```
The same, should be mocked for right answer as well.

maZahaca commented on tests/EcbCurrencyRateProviderTest.php:
```diff
@@ -157,5 +159,34 @@ public function testEcbCurrencyRateProviderGetRatesYesterday()
+
+        try {
+            $this->currencyRateProvider->getRates(array_values($currencies));
+        } catch (\Exception $e) {
```
I guess that would be more useful:

```
/**
* @expectedException BadXMLQueryException            
* @expectedExceptionMessageRegExp #No rates available for ....-..-.. date with provider ecb#
*/
```

remedge replied:
fixed.

maZahaca commented on tests/YahooCurrencyRateProviderTest.php:
```diff
@@ -0,0 +1,185 @@
+        $currencyManager = $this->getMock('\\RedCode\\Currency\\ICurrencyManager');
+        $currencyRateManager = $this->getMock('\\RedCode\\Currency\\Rate\\ICurrencyRateManager');
+
+        $xmlLoader = $this->getMock('\\RedCode\\Currency\\Rate\\XML\\XMLLoader');
```
Should be covered for right answers as well

maZahaca commented:
Check this issues https://scrutinizer-ci.com/g/redco/redcode-currency-rate/inspections/7e2cb2fc-be7a-42db-9f04-887fa6dcd3e1/issues/

remedge committed:
improve tests.

remedge committed:
use libxml_use_internal_errors to catch errors.

remedge committed:
update tests again.

remedge committed:
keep fighting.

remedge committed:
simplify exception.

remedge committed:
fix scrutinizer duplication issues.

remedge commented:
can't figuring out what scrutinizer means on duplication.

And i fixed all other issues you pointed.

maZahaca commented on tests/YahooCurrencyRateProviderTest.php:
```diff
@@ -0,0 +1,210 @@
+        $currencyRateManager = $this->getMock('\\RedCode\\Currency\\Rate\\ICurrencyRateManager');
+
+        $quote = (object)['quote' => []];
+        $xmlResponse = (object)['results' => $quote];
```
You need here to add real xml response from yahoo provider, and try on it

maZahaca replied:
You can add methods in the tests, which return mocked XMLLoader with 
- correct response and 
- incorrect response 

And use them in places where that is needed.
Do the same for Ecb porvider

maZahaca replied:
So there is no way in you provider tests to use `new XMLLoader`

maZahaca commented on tests/ExceptionTest.php:
```diff
@@ -28,20 +50,37 @@ public function testRateNotFoundException()
+
+    public function testBadXMLQueryException()
+    {
+
```
spaces to be deleted

maZahaca commented:
> scrutinizer means on duplication

That's okay about this

maZahaca commented:
Last changes about mocking of XMLLoaded are needed.
Other stuff is pretty good! :+1:

remedge committed:
refactoring tests.

remedge commented:
Seems it's done. Tests are been refactored little bit.

maZahaca commented on src/RedCode/Currency/Rate/Exception/BadXMLQueryException.php:
```diff
@@ -0,0 +1,33 @@
+    public function __construct($query, ICurrencyRateProvider $provider)
+    {
+        $this->query = $query;
+        $this->provider = $provider;
```
It's better to hide all properties which are possible to hide.

So the solution using of base class constructor `parent::__construct($provider)` is more preferable.
This will also add strict dependency on `$provider` for base class.

In the base class change `protected $provider;` to `private $provider;`
So the base class will be like this:

``` PHP
abstract class BaseProviderException extends \Exception
{
    /**
     * @var \RedCode\Currency\Rate\Provider\ICurrencyRateProvider
     */
    private $provider;

    public function __construct(ICurrencyRateProvider $provider)
    {
        $this->provider = $provider;
    }

    /**
     * @return ICurrencyRateProvider
     */
    public function getProvider()
    {
        return $this->provider;
    }
}
```

Probably, the two issues of [scrutinizer-ci](https://scrutinizer-ci.com/g/redco/redcode-currency-rate/inspections/946189de-faa9-468d-a507-dd3653c03d31) are about that, but not sure.

remedge committed:
update BaseProviderException.

use parent::__construct($provider).

remedge commented:
Unfortunately, scrutinizer continues to complain even after this fix

maZahaca commented:
Now it's perfect, thanks!
